File: baseunits/modules/MyReadingManga.pas
unit MyReadingManga;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML;

implementation

uses httpsendthread;

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation; var Page: Integer; const WorkPtr: Integer;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL) then begin
    Result := NO_ERROR;
    Page := StrToIntDef(XPathString('//*[contains(@class,"archive-pagination")]/ul/li[last()-1]', MangaInfo.FHTTP.Document), 1);
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo.FHTTP.GET(Module.RootURL + '/page/' + IncStr(AURL) + '/') then
  begin
    Result := NO_ERROR;
    XPathHREFAll('//h2[@class="entry-title"]/a', MangaInfo.FHTTP.Document, ALinks, ANames);
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := MaybeFillHost(Module.RootURL, AURL);
    if GET(url) then begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do try
        title := XPathString('//*[contains(@class,"entry-content")]/h2');
        if title = '' then title := Trim(XPathString('//*[contains(@class,"entry-content")]/p[starts-with(.,"Title")]/substring-after(.,":")'));
        if title = '' then title := XPathString('//*[contains(@class,"entry-content")]/p[1]/strong');
        if title = '' then title := XPathString('//h1[@class="entry-title"]');
        genres := XPathString('//header[@class="entry-header"]/string-join(./p[position()>1]//a,", ")');
        authors := Trim(XPathString('//*[contains(@class,"entry-content")]/p[starts-with(.,"Author")]/substring-after(.,":")'));
        chapterLinks.Add(url);
        chapterName.Add(title);
        for v in XPath('//*[contains(@class,"entry-pagination")]/a') do
        begin
          chapterLinks.Add(v.toNode.getAttribute('href'));
          chapterName.Add(title + ' - ' + v.toString);
        end;
        if chapterName.Count > 1 then
          chapterName[0] := chapterName[0] + ' - 1';
      finally
        Free;
      end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GET(MaybeFillHost(Module.RootURL, AURL)) then
    begin
      Result := True;
      XPathStringAll('//*[contains(@class,"entry-content")]//div//img/@src', Document, PageLinks);
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do
  begin
    Website := 'MyReadingManga';
    RootURL := 'https://myreadingmanga.info';
    Category := 'H-Sites';
    SortedList := True;
    OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/MyReadingManga.lua
function GetInfo()
	local u = MaybeFillHost(module.RootURL, url)

	if not http.get(u) then return net_problem end

	local x = TXQuery.Create(http.Document)
	mangainfo.Title  = x.XPathString("//h1[@class='entry-title']")
	mangainfo.Genres = x.XPathString('//header[@class="entry-header"]/string-join(./p[position()>1]//a,", ")')
	mangainfo.Status = MangaInfoStatusIfPos(x.XPathString('//*[@class="entry-terms" and contains(., "Status")]/a'))

	mangainfo.ChapterLinks.Add(mangainfo.url)
	mangainfo.ChapterNames.Add(mangainfo.title)

	local v = x.XPath('//*[contains(@class,"entry-pagination")]/a')
	for i = 1, v.Count do
		local v1 = v.Get(i)
		if string.match(v1.toString, '^Next') == nil then
			mangainfo.ChapterLinks.Add(v1.getAttribute('href'));
			mangainfo.ChapterNames.Add(mangainfo.Title .. ' - ' .. v1.toString)
		end
	end
	if mangainfo.ChapterNames.Count > 1 then
		mangainfo.ChapterNames[0] = mangainfo.ChapterNames[0]
	end
end

function GetPageNumber()
	task.PageLinks.Clear()
	if http.get(MaybeFillHost(module.RootURL, url)) then
		local x = TXQuery.Create(http.Document)
		x.XPathStringAll('//*[contains(@class,"entry-content")]//img/@data-src', task.PageLinks)
		if task.PageLinks.Count == 0 then
			x.XPathStringAll('//div[@class="separator" and @style]//img/@data-src', task.PageLinks)
		end
	else
		return false
	end
	return true
end

function GetDirectoryPageNumber()
	if http.get(module.RootURL) then
		local x = TXQuery.Create(http.Document)
		page = tonumber(x.XPathString('//*[contains(@class,"archive-pagination")]/ul/li[last()-1]'))
		if page == nil then page = 1 end
		return no_error
	else
		return net_problem
	end
end

function GetNameAndLink()
	if http.get(module.rooturl .. '/page/' .. IncStr(url) .. '/') then
		local x = TXQuery.Create(http.Document)
		x.XPathHREFAll('//h2[@class="entry-title"]/a', links, names)
		return no_error
	else
		return net_problem
	end
end

function Init()
	local m = NewModule()
	m.Website = 'MyReadingManga'
	m.RootURL = 'https://myreadingmanga.info'
	m.Category = 'H-Sites'
	m.OnGetInfo='GetInfo'
	m.OnGetPageNumber='GetPageNumber'
	m.OnGetNameAndLink='GetNameAndLink'
	m.OnGetDirectoryPageNumber = 'GetDirectoryPageNumber'
	m.SortedList = true
end
